diff --git src/options.ml src/options.ml
index fbf9d6a..7b4913b 100644
--- src/options.ml
+++ src/options.ml
@@ -94,6 +94,7 @@ let no_pervasives = ref false
 let files_to_load_ = ref []
 let show_time_ = ref false
 let locs_files = ref []
+let default_locs = ref false
 let explain_vc = ref false
 let locs_table = Hashtbl.create 97
 
@@ -236,6 +237,7 @@ Typing/Annotations/VCG options:
   --total            total correctness
   --explain          outputs explanations for VCs in file.xpl
   --locs f           reads source locations from file f
+  --default-locs     reads source locations from file basename.loc
   --phantom <name>   declare <name> as a phantom type
 
 VC transformation options:
@@ -604,6 +606,9 @@ let files =
     | ("-locs" | "--locs") :: s :: args ->
 	locs_files := s :: !locs_files;
 	parse args
+    | ("-default-locs" | "--default-locs" ) :: args ->
+          default_locs := true;
+          parse args
     | ("-phantom" | "--phantom") :: s :: args ->
 	Hashtbl.add phantom_types s ();
 	parse args
@@ -732,6 +737,11 @@ let if_debug_2 f x y = if debug then f x y
 let if_debug_3 f x y z = if debug then f x y z
 
 let () =
+  if !default_locs then
+    locs_files :=
+       (List.map (fun s ->
+          Filename.chop_extension s ^ ".loc" ) files) @ !locs_files;
+
   List.iter
     (fun f ->
        let l = Rc.from_file f in
